Blizzard.IO.RabbitMQ/NetqRabbitConsumer.cs

```diff
@@ -0,0 +1,92 @@
+using EasyNetQ.Topology;
+using Microsoft.Extensions.Logging;
+
+namespace Blizzard.IO.RabbitMQ
+{
+    public class NetqRabbitConsumer<TData> : IConsumer<TData>, IConsumerWithMetadata<TData, RabbitMessageProperties>
+    {
+        private readonly IBus _netqBus;
+        private readonly IDeserializer<TData> _deserializer;
+        private readonly IAbstractTypeDeserializer<TData> _concreteTypeDeserializer;
+        private readonly IQueue _netqQueue;
+        private readonly IConverter<MessageProperties, RabbitMessageProperties> _converter;
+        private readonly ILogger _logger;
+        
+        private IDisposable _consumeHandler;
+        private readonly Func<byte[], MessageProperties, TData> _dataExtractionStrategy;
+
+        public event Action<TData> MessageReceived;
+        public event Action<TData, RabbitMessageProperties> MessageWithMetadataReceived;
+
+        protected NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, ILoggerFactory loggerFactory,
+            IConverter<MessageProperties, RabbitMessageProperties> coverter)
+        {
+            _netqBus = netqBus;
+            _netqQueue = new Queue(sourceQueue.Name, sourceQueue.Exclusive);
+            _logger = loggerFactory.CreateLogger(nameof(NetqRabbitConsumer<TData>));
+            _converter = coverter ?? new RabbitPropertiesConverter();
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IDeserializer<TData> deserializer, 
+            ILoggerFactory loggerFactory, IConverter<MessageProperties, RabbitMessageProperties> coverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, coverter)
+        {
+            _deserializer = deserializer;
+            _dataExtractionStrategy += ExtractData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} data");
+        }
+
+        public NetqRabbitConsumer(IBus netqBus, RabbitQueue sourceQueue, IAbstractTypeDeserializer<TData> concreteTypeDeserializer,
+            ILoggerFactory loggerFactory, IConverter<MessageProperties, RabbitMessageProperties> coverter = null)
+            : this(netqBus, sourceQueue, loggerFactory, coverter)
+        {
+            _concreteTypeDeserializer = concreteTypeDeserializer;
+            _dataExtractionStrategy += ExtractAbstractData;
+            _logger.LogDebug($"Initialized consumer from queue: {_netqQueue.Name}," +
+                $" expecting {typeof(TData)} derieved data");
+        }
+
+        public void Start()
+        {
+            _consumeHandler = _netqBus.Advanced.Consume(_netqQueue, (messageBytes, messageProperties, messageIndo) =>
+            {
+                TData data = _dataExtractionStrategy.Invoke(messageBytes, messageProperties);
+                RabbitMessageProperties properties = _converter.Convert(messageProperties);
+                MessageReceived(data);
+                MessageWithMetadataReceived(data, properties);
```

Comment:
Use the .?Invoke() function on the events you are going to get exception if you did not subscribed to both of the events

Reply:
f